### PR TITLE
make jsonrpc api namespaces extensible

### PR DESCRIPTION
**Pull request by yihuang**
## Description

Closes: #1052, #1037

An example of usage: https://github.com/crypto-org-chain/cronos/pull/432/files#diff-7acb7332ad576a49e7685db72282959006f156c06a708828d8b0b187e311ffe1R32

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

**Review by fedekunze — COMMENTED** [review]
I like the approach but ideally, you shouldn't be able to override existing namespaces, just add new ones. Can you update the implementation to support that?

File: rpc/apis.go [hunk 1]
```diff
@@ -35,88 +35,105 @@ const (
 	apiVersion = "1.0"
 )
 
+// APICreator creates the json-rpc api implementations.
+type APICreator = func(*server.Context, client.Context, *rpcclient.WSClient) []rpc.API
+
+// APICreators defines namespaced api creators, can be extended by external modules.
```
Comment (crypto-facs) [inline rpc/apis.go:41]:
```suggestion
// APICreators defines namespaces api creators, can be extended by external modules.
```

**Comment by yihuang** [comment]
> I like the approach but ideally you shouldn't be able to override existing namespaces, just add new ones. Can you update the implementation to support that?

good idea, added a method to register extra namespace.

File: rpc/apis.go [hunk 2]
```diff
@@ -35,94 +37,124 @@ const (
...
+			apis = append(apis, creator(ctx, clientCtx, tmWSClient)...)
+		} else {
+			ctx.Logger.Error("invalid namespace value", "namespace", ns)
 		}
 	}
 
 	return apis
 }
 
+// RegisterAPINamespace registers an api namespace with the api creator.
```
Comment (fedekunze) [inline rpc/apis.go:149]:
```suggestion
// RegisterAPINamespace registers a new API namespace with the API creator.
// This function fails if the namespace is already registered.
```

File: rpc/apis.go [hunk 3]
```diff
@@ -35,94 +37,124 @@ const (
...
+			}
+		},
+	}
+}
+
+// GetRPCAPIs returns the list of all APIs
+func GetRPCAPIs(ctx *server.Context, clientCtx client.Context, tmWSClient *rpcclient.WSClient, selectedAPIs []string) []rpc.API {
+	var apis []rpc.API
+	// remove duplicates
+	selectedAPIs = unique(selectedAPIs)
```
Comment (fedekunze) [inline rpc/apis.go:136]: we can remove the `unique` function, it should be checked by the config validation IIRC.

```suggestion
```

File: rpc/apis.go [hunk 4]
```diff
@@ -35,94 +37,124 @@ const (
...
+// RegisterAPINamespace registers an api namespace with the api creator.
+func RegisterAPINamespace(ns string, creator APICreator) error {
+	if _, ok := apiCreators[ns]; ok {
+		return fmt.Errorf("duplicated api namespace %s", ns)
+	}
+	apiCreators[ns] = creator
+	return nil
+}
+
 func unique(intSlice []string) []string {
```
Comment (fedekunze) [inline rpc/apis.go:158]: let's remove this function

**Review by fedekunze — APPROVED** [review]
ACK, minor comments

**Pull request merged**